internal/controller/pkg/transaction/constructor.go

```diff
@@ -0,0 +1,143 @@
+*/
+
+// Package transaction implements the Transaction controller.
+package transaction
+
+import (
+	"strings"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	"github.com/crossplane/crossplane-runtime/v2/pkg/conditions"
+	"github.com/crossplane/crossplane-runtime/v2/pkg/errors"
+	"github.com/crossplane/crossplane-runtime/v2/pkg/event"
+	"github.com/crossplane/crossplane-runtime/v2/pkg/logging"
+
+	"github.com/crossplane/crossplane/v2/apis/pkg/v1alpha1"
+	"github.com/crossplane/crossplane/v2/apis/pkg/v1beta1"
+	"github.com/crossplane/crossplane/v2/internal/controller/pkg/controller"
+	"github.com/crossplane/crossplane/v2/internal/controller/pkg/revision"
+	"github.com/crossplane/crossplane/v2/internal/xpkg"
+	"github.com/crossplane/crossplane/v2/internal/xpkg/dependency"
+)
+
+// ReconcilerOption configures a Reconciler.
+type ReconcilerOption func(*Reconciler)
+
+// WithLogger specifies how the Reconciler should log messages.
+func WithLogger(log logging.Logger) ReconcilerOption {
+	return func(r *Reconciler) {
+		r.log = log
+	}
+}
+
+// WithRecorder specifies how the Reconciler should record Kubernetes events.
+func WithRecorder(er event.Recorder) ReconcilerOption {
+	return func(r *Reconciler) {
+		r.record = er
+	}
+}
+
+// WithLockManager specifies how the Reconciler should manage Lock access.
+func WithLockManager(lm LockManager) ReconcilerOption {
+	return func(r *Reconciler) {
+		r.lock = lm
+	}
+}
+
+// WithDependencySolver specifies how the Reconciler should resolve dependencies.
+func WithDependencySolver(ds DependencySolver) ReconcilerOption {
+	return func(r *Reconciler) {
+		r.solver = ds
+	}
+}
+
+// WithValidator specifies how the Reconciler should validate Transactions.
+func WithValidator(v Validator) ReconcilerOption {
+	return func(r *Reconciler) {
+		r.validator = v
+	}
+}
+
+// WithValidators specifies a chain of validators for the Reconciler.
+// Validators run in sequence, failing fast on the first error.
+func WithValidators(validators ...Validator) ReconcilerOption {
+	return func(r *Reconciler) {
+		r.validator = ValidatorChain(validators)
+	}
+}
+
+// WithInstaller specifies how the Reconciler should install packages.
+func WithInstaller(i Installer) ReconcilerOption {
+	return func(r *Reconciler) {
+		r.installer = i
+	}
+}
+
+// NewReconciler returns a Reconciler of Transactions.
+func NewReconciler(kube client.Client, pkg xpkg.Client, opts ...ReconcilerOption) *Reconciler {
+	r := &Reconciler{
+		kube:       kube,
+		pkg:        pkg,
+		log:        logging.NewNopLogger(),
+		record:     event.NewNopRecorder(),
+		conditions: conditions.ObservedGenerationPropagationManager{},
+		lock:       NewAtomicLockManager(kube),
+		solver:     dependency.NewTighteningConstraintSolver(pkg),
+		validator: ValidatorChain{
+			// TODO(negz): Validate RBAC, etc.
```

Comment:
Just saw this while putting together a demo.

```
    Message:               cannot establish control of package objects: bucketnotifications.s3.aws.m.upbound.io is already controlled by ProviderRevision arbitrary-name-provider-s3-06a50063a67d (UID a2c61d04-fc42-42e5-8ac3-025f8eafdad3)
````

It happened during the install phase, i.e. past the point of no return. So validating that we could establish control of all objects, before we actually try, would be a good idea.